src/EditorFeatures/CSharpTest/Diagnostics/GenerateMethod/GenerateDeconstructMethodTests.cs

```diff
@@ -0,0 +1,300 @@
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        var (x, y) = this;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionDeclaration_BuiltinType()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        (int x, int y) = [|1|];
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionAssignment()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        int x, y;
+        (x, y) = [|this|];
+    }
+}",
+@"using System;
+
+class Class
+{
+    private void Deconstruct(out int x, out int y)
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        int x, y;
+        (x, y) = this;
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionAssignment_Nested()
+        {
+            // We only offer a fix for non-nested deconstruction, at the moment
+            await TestMissingInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        int x, y, z;
+        ((x, y), z) = ([|this|], 0);
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestDeconstructionAssignment_Array()
+        {
+            await TestMissingInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        int x, y;
+        (x, y) = [|new[] { this }|];
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestSimpleDeconstructionForeach()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Class
+{
+    void Method()
+    {
+        foreach ((int x, int y) in new[] { [|this|] }) { }
+    }
+}",
+@"using System;
+
+class Class
+{
+    private void Deconstruct(out int x, out int y)
+    {
+        throw new NotImplementedException();
+    }
+
+    void Method()
+    {
+        foreach ((int x, int y) in new[] { this }) { }
+    }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateMethod)]
+        public async Task TestSimpleDeconstructionForeach_AnotherType()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Class
+{
+    void Method(D d)
+    {
+        foreach ((int x, int y) in new[] { [|d|] }) { }
+    }
+}
+class D
+{
+}",
+@"using System;
+
+class Class
+{
+    void Method(D d)
+    {
+        foreach ((int x, int y) in new[] { d }) { }
+    }
+}
+class D
+{
+    internal void Deconstruct(out int x, out int y)
+    {
+        throw new NotImplementedException();
+    }
+}");
+        }
+    }
```

Comment:
The tests use `TestInRegularAndScriptAsync` to test in script context, but since all of them operate within a class there isn't much difference between script and non-script. 
Seems unnecessary to spent cycles running the tests twice.
Instead we should have a couple of tests that invoke the code fix in a top-level method and a top-level statement.

Reply:
I looked at this, but it turns out there isn't even a test helper for just testing "regular".
That seems to be the approach most or all IDE tests take, even if they involve a class declaration.